### PR TITLE
Add libddwaf metrics

### PR DESCRIPTION
**Pull request by lloeki**
Add libddwaf telemetry metrics and metadata as described by the corresponding AppSec RFC.

- Ruleset version must be sent with each trace to correlate with libddwaf runs (whether they trigger security events or not)
- As much as possible is sent only on the first trace. This 
- A wrapping class around `Datadog::AppSec::WAF::Context` is introduced, to track and report execution time (overall + internal) and timeout occurence

Spec for `Processor` has been updated, while `Rack::RequestMiddleware` and `Watcher` paths are exercised via System Tests (I wish these could report code coverage and fold it into Codecov):

```
runner           | XPASSED tests/appsec/waf/test_reports.py::Test_Monitoring::test_waf_monitoring - Expected to fail, but all is ok
runner           | XPASSED tests/appsec/waf/test_reports.py::Test_Monitoring::test_waf_monitoring_once - Expected to fail, but all is ok
```

### CODE REVIEW

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [inline lib/datadog/appsec/contrib/rack/request_middleware.rb:79]: Minor: There's a potential here for a weird interaction. If the customers start with tracing disabled (or for some reason have a special request that skips tracing), we'll mark the request as first, but never actually add this information (due to the `return unless active_trace`).

Should we perhaps only set that flag _after_ sending it on a trace?

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [inline lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: Minor: This is a bit weird since it can override an explicit customer `reject!`. Should this check first if `sampled?` is set before calling the `keep!`?

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [inline lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: ‼️ ‼️ Important: I think there's a bug here -- if `@processor.ready? == false` in line 24, then this will be called with `nil` and the `add_waf_runtime_tags(nil)` will break the app.

If my assessment is correct, in addition to fixing it, I would also request adding test coverage for this :)

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [inline lib/datadog/appsec/contrib/rack/request_middleware.rb:100]: Minor: Would it make sense to have helper methods that return time in the intended unit (it appears to me this is _microseconds_, but I may be wrong...?)

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [inline lib/datadog/appsec/processor.rb:39]: I suggest being explicit with the units here:

```suggestion
          start_ns = Core::Utils::Time.get_time(:nanosecond)

          ret, res = @context.run(*args)

          stop_ns = Core::Utils::Time.get_time(:nanosecond)

          @time_ns += res.total_runtime # <-- I'm not even sure what unit this is in :(
          @time_ext_ns += (stop_ns - start_ns)
          @timeouts += 1 if res.timeout
```

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [inline lib/datadog/appsec/processor.rb:24]: This is a weird break in the abstraction... Have you considered passing the `handle` directly when creating the new `Context`?

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [inline lib/datadog/appsec/processor.rb:20]: Minor: In the spirit of self-documenting code, have you considered perhaps naming this `ContextWrapper` and removing the comment? ;)

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [inline lib/datadog/appsec/processor.rb:51]: Minor: Should `@addresses` also be initialized here to `nil`?

**Review by ivoanjo — COMMENTED** [review]
Left a few notes!

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (marcotc) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: I believe if this span represents a security alert, it would override sampling, even if the user explicit doesn't want it retained, is that correct @lloeki?

I can see this going in two ways, depending on what the Appsec team wants here: one might not know they a security issue in their endpoint and not informing the user because they chose to not perform tracing collection could be seem as incorrect. But at the same time, if the endpoint is clearly inconsequential (e.g. a health-check endpoint, or internal-only route) then I can see the argument to respect `keep!` and `reject!`.

If you want to check for explicit user sampling, I'd suggest adding a new method to `TraceOperation`:
```ruby
      # The user has explicitly set the tracing decision for this operation.
      def user_sampling?
        sampling_priority == Sampling::Ext::Priority::USER_KEEP || sampling_priority == Sampling::Ext::Priority::USER_REJECT
      end
```
I don't think there's a good way to know this today with the currently exposed API in `TraceOperation`, so a new method would work.

File: lib/datadog/appsec/processor.rb
Comment (marcotc) [reply lib/datadog/appsec/processor.rb:20]: I bet removing the comment will make Rubocop cry :/, but the renaming is a good suggestion.

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:24]: You mean using `send` here? I was annoyed by that too and considered it.

The abstraction goal here with `Processor` and `Processor::Context` is to hide away `Datadog::AppSec::WAF`, so exposing the handle (which is a `Datadog::AppSec::WAF::Handle`) runs against that.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: Good question and astute analysis. I need to confirm with the team what the expectation is, although I think it is expected to be retained in spite of being rejected by the user.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:79]: As it stands, if the app starts with `Tracing` disabled there is no trace that would ever be sent, therefore it becomes immaterial. So AppSec would presumably work but be unable to send any data to the backend. That is, until traces become a proper `Core` data bus independent of `Tracing`.

That might change with remote configuration so I understand the logic here.

> for some reason have a special request that skips tracing

Is that possible?

Still, I will see if I can come up with something.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:82]: Needed, probably not. Required by RFC, yes: it was decided that this tag needs to be sent irrespective of any other tag that would achieve the same goal, IIRC to cater for the current lack of actual telemetry or consistent tags across languages (I don't know the exact details that led to that decision though).

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:110]: Yes it does, although only for data transport. I'm trying to keep `AppSec` as detached as possible from `Tracing`. Ideally it should be able to fully execute with tracing disabled, and only not report any data because of the lack of trace transport (FYI there is a blocking feature, currently unused: no rules in the current recommended ruleset is set to return a blocking decision. With tracing disabled AppSec should still be able to block).

I must admit I did not exercise that case lately and there may be gaps, but that's the general idea. Also event propagation in #1969 may pose an issue depending on the chosen implementation.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:59]: Ah, I'm very used to using bangs, so in my mind `request!` is immediately read as "mark request", which in this context is exactly equivalent to `this request has been performed`.

I understand that it may sound confusing though (probably I did too much Perl in my life and got too used to non-ambiguous inference from context).

Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:59]: Also, this may tie in together with https://github.com/DataDog/dd-trace-rb/pull/1987#discussion_r864669076 as it may not be that the request has been processed that may matter but the fact that tags have been transmitted at least once.

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:110]: Also I'm not sure but I think `&& Datadog::Tracing.respond_to?(:active_trace)` is not needed anymore. It was a carryover from `Datadog.active_root_span` which may not have been there in all cases but I believe `defined?(Datadog::Tracing)` is sufficient now.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:70]: Hmm, I wish there were [RWMutex](https://pkg.go.dev/sync#RWMutex) in Ruby: readers don't lock between each other as with a simpler `Mutex`, only [read vs write does](https://stackoverflow.com/a/19168242):

> [RWMutex](https://pkg.go.dev/sync#RWMutex) is an extension of [Mutex](https://pkg.go.dev/sync#Mutex) and adds two methods: [RLock](https://pkg.go.dev/sync#RWMutex.RLock) and [RUnlock](https://pkg.go.dev/sync#RWMutex.RUnlock). There are a few points that are important to note about [RWMutex](https://pkg.go.dev/sync#RWMutex):
>
>    [RLock](https://pkg.go.dev/sync#RWMutex.RLock) is a shared read lock. When a lock is taken with it, other threads* can also take their own lock with [RLock](https://pkg.go.dev/sync#RWMutex.RLock). This means multiple threads* can read at the same time. It's semi-exclusive.
>
> -  If the mutex is read locked, a call to [Lock](https://pkg.go.dev/sync#Mutex.Lock) is blocked**. If one or more readers hold a lock, you cannot write.
>
>   - If the mutex is write locked (with [Lock](https://pkg.go.dev/sync#Mutex.Lock)), [RLock](https://pkg.go.dev/sync#RWMutex.RLock) will block**.
>
> A good way to think about it is [RWMutex](https://pkg.go.dev/sync#RWMutex) is a [Mutex](https://pkg.go.dev/sync#Mutex) with a reader counter. [RLock](https://pkg.go.dev/sync#RWMutex.RLock) increments the counter while [RUnlock](https://pkg.go.dev/sync#RWMutex.RUnlock) decrements it. A call to [Lock](https://pkg.go.dev/sync#Mutex.Lock) will block as long as that counter is > 0.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: Answer: currently undefined. There are pros and cons, but it seems like the user having the final word weighs in big.

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:39]: `Core::Utils::Time.get_time(:nanosecond)`: Cool, I didn't knwo this was a thing!

Good point, although I like having unit-typed values+type checking instead of using hungarian notation (which is merely documentation and, although valuable, does not prevent mistakes by itself).

I need to remember to change the `attr_readers` too!

Comment (lloeki) [reply lib/datadog/appsec/processor.rb:39]: (and yes it's nanoseconds)

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:100]: I like having unit-typed values+type checking instead of using hungarian notation, I venture conversion methods in `Core::Utils::Time` could be a nice way to gradually annotate the code with unit information, and eventually have them return proper unit types, which would then be able to be type-checked.

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [reply lib/datadog/appsec/processor.rb:24]: Ah, no, I meant actually passing in the fully-built context from the `Processor`. E.g. instead of

```ruby
      def new_context
        Context.new(self)
      end
```

why not do

```ruby
      def new_context
        Context.new(Datadog::AppSec::WAF::Context.new(@handle))
      end
```

...that way the handle never needs to go out of the processor, and the `Processor::Context` doesn't actually need to care about where the object it wraps comes from.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:79]: > As it stands, if the app starts with Tracing disabled there is no trace that would ever be sent, therefore it becomes immaterial. So AppSec would presumably work but be unable to send any data to the backend. That is, until traces become a proper Core data bus independent of Tracing.

Right. My point is that in such a case, we'll still mark the request as not being the first one, and in such a case the special "first request" metadata would never be sent.

> > for some reason have a special request that skips tracing
>
> Is that possible?

I guess it could if customers enable appsec and its instrumentation, then do a request, and then enable tracing and its instrumentation.

Probably both of the cases I'm pointing out here are quite corner cases, but since the solution is quite simple -- don't mark a request as first until we actually send the special "first request" metadata I think it may be worth fixing.

(The change I propose above probably makes `@first_request` a bit misnamed, so probably it would become `@appsec_info_sent_once` or something like that?)

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:82]: I don't quite agree with that reasoning -- it sounds like a decision that was taken without context of what we already had.

But, I won't block the PR on it either. If you still want it, I'll disagree and commit.

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:110]: Interesting -- so would it be possible and a supported configuration to run AppSec with tracing disabled?

That would indeed change my observation, as in that case it would mean that even with tracing we'd want this class to do things.

Is that something we already support, or something that we're going to add later?

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:20]: I guess it would be *exact* but I don't feel it would be *useful*: it would only expose internals from an abstraction that is supposed to hide what is happening behind the scenes and augment it with additional features.

This makes me think it's in fact the comment that is incorrect now.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:70]: +1 It sucks that the Ruby stdlib has so little in concurrency utilities. It often pushes people to roll their own which is even worse.

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [reply lib/datadog/appsec/processor.rb:39]: +1 I also prefer using types for this. When we're using primitives instead, I always like to annoy people to document via code :)

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:110]: We don't support it today but it was a requirement early on, back when the transport was separate the trace API and used a dedicated AppSec API.

Since I already wrote the code in a very isolated way from tracing, I find it both easier and beneficial to keep up with this separation of concerns as it would be hard to retrofit later on if coupling creeps in.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:70]: Maybe we can do a gem that would implement that, and contribute it to upstream Ruby? ;)

Anyway, this has been refactored and comments added.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:79]: Refactored into something like that.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:82]: > If you still want it

I don't really have a say here: this was decided outside of my sphere of responsibility.

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:51]: One is not really supposed to query `addresses` if `create_waf_handle` failed, but for the sake of consistency, sure.

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:24]: Hmm, I understand. Then `Datadog::AppSec::WAF::Context` becomes part of the public interface of `Processor::Context` as an argument to `new`. I feel like that would make it a leaky abstraction.

File: lib/datadog/appsec/processor.rb
Comment (lloeki) [reply lib/datadog/appsec/processor.rb:39]: Done

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb
Comment (marcotc) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:110]: If the goal is to move a Tracing-independent Appsec implementation, than I'm ok with the changes as is.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: Correct. It can also be the case if `@processor.new_context` fails.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: Ah I was confused: this only happens on the first request, in order to send telemetry, irrespective of attacks. In this case it is mandatory to send it.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:100]: Hmm, without types it gets unwieldy, with potentially a lot of X_to_Y conversions (here, ns_to_us). I added a comment though, which, in addition to the `_ns` suffix, makes things a whole lot clearer.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: Okay, so I tried adding a spec for that and turns out it's a rabbit hole and derailing the PR. I'd rather address the spec aspect (especially those that run under contrib and appraisals) more generally by landing a proper architecture for that instead of hacking things around, which would be worse.

Comment (marcotc) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: > I'd rather address the spec aspect (especially those that run under contrib and appraisals) more generally by landing a proper architecture for that instead of hacking things around, which would be worse.

Even without specs, I think we should still guard around this case, to avoid calling `add_waf_runtime_tags(nil)`

**Review by marcotc — CHANGES_REQUESTED** [review]

**Comment by marcotc** [comment]
Looks good, besides the one comment I've just left. 👍

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: Haha, but I did! Just not at this call site in `ensure` but within `add_waf_runtime_tags`:

```
          def add_waf_runtime_tags(context)
            return unless active_trace
            return unless context            # <---- there
```

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb]: We could even call such a gem `concurrent-something` ;) ;)

File: lib/datadog/appsec/processor.rb
Comment (ivoanjo) [reply lib/datadog/appsec/processor.rb:24]: Fair enough... I guess?

I think having to use the `send(:handle)` here is already breaking the abstraction anyway but seeing as this already quite localized, and it's rather small, let's keep it as-is, and we can always revisit it later as the codebase evolves. Marking as resolved 😄

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: 👍  I would actually suggest going one step further and also doing it in the `ensure` as well, because this subtlety might not be obvious to the next person that touches this code, which may add another call to another method and forget to do the check.

`add_waf_runtime_tags(context) if context` would make it clear that `context` can be `nil`.

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (ivoanjo) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:92]: @marcotc would it make sense for `TraceOperation` to emit a log warning when a trace has gone from `USER_REJECT` to `USER_KEEP`? At least in that case customers would have visibility on why such a weird thing happened.

**Review by ivoanjo — APPROVED** [review]
LGTM 👍

File: lib/datadog/appsec/contrib/rack/request_middleware.rb
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/request_middleware.rb:58]: Fair enough

**Pull request merged**
